Add stack trace logging to the grpc-recovery middleware

rdner opened the pull request:
Add logging and stacktrace on panic recovery
    
**What**
Now every time the grpc recovery middleware recovers from panic it logs the error message using logrus including the stacktrace.

**Why**
It was hard to troubleshoot panics because of the absence of the logs.

Also:
- I had to fix some tests to make them passing (had race conditions and false checks)
- I had to update a dependency to build the project (it was missing method)
- I changed the package names according to the common convention (directory=package name).

rdner committed:
Fix project build

Was not able to build the project because of:

`grpc/tracing.go:30:8: undefined:
otgrpc.OpenTracingStreamServerInterceptor`

Updating the dependency helped.

rdner committed:
Change package names according to the convention

By the convention a package name should match a directory name a go
file in.

rdner committed:
Improve logging in test

It was not clear why the test was failing.

rdner committed:
Restore the logger output after a test

**What**
Restore a logrus output to stdout after each test that sets it to a
buffer

**Why**
If we don't do that one test can interfere with another that check the
logger output.

rdner committed:
Fix the metadatalogging test

**What**
- Switched off colors for logging
- Add a time.Sleep before trying to access the server

**Why**
- It tried to match a colored string and sometime the logger output a
colored entry
- It tried to make a request before the server was running

All these problems were reproducible only occasionally, sometimes the
test was passing.

rdner committed:
Add logging and stacktrace on panic recovery

**What**
Now every time the grpc recovery middleware recovers from panic it
logs the error message using logrus including the stacktrace.

**Why**
It was hard to troubleshoot panics because of the absence of the logs.

rdner commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Add stack trace logging to the grpc-recovery middleware](https://trello.com/c/GkDeZIB9/262-add-stack-trace-logging-to-the-grpc-recovery-middleware)

LucasRoesler commented on grpc/recovery_test.go:
```diff
@@ -1,28 +1,43 @@
 		defer cancel()
+
 		go ListenAndServe(ctx, ":3005", srv)
+		time.Sleep(time.Second)
```
Add a comment here so that this is not deleted in the future

rdner replied:
Added the comment and `time.Sleep` in other places where it's necessary. Also, decreased the time to 100 milliseconds which should be enough.

LucasRoesler commented on grpc/recovery_test.go:
```diff
@@ -1,28 +1,43 @@
 	It("should recover from panic when recovery option is set", func() {
+		buf := &bytes.Buffer{}
+		logrus.SetOutput(buf)
+		defer func() {
```
I think it is probably safe to standardize this into a function that  always reset to info level and stdout, it will not really break anything, make the test environment more predictable and reduce the amount of boiler plate in the tests.

rdner replied:
I extracted the logging manipulation into `goserver/test/utils.go` also disabled logs in both `grpc` and `http` test suites except the buffering cases where we check the logs content in tests.

However, there are still two log entries on the console and they happen because some of the servers are still running after I run restore of the `DiscardLogging`. Probably the one in metrics `go goserver.ListenAndServeMetricsAndHealth(":8080", nil)` because it does not accept a context to cancel.

LucasRoesler replied:
Sounds like you found another trello ticket :)

rdner committed:
Refactor unit tests and add comments

**What**
- Added an util package for suppressing and buffering logging in order
to reduce the boiler plate code in unit tests.
- Add time.Sleep everywhere necessary, commented why

**Why**
The unit tests used the same code to manipulate logging which was
hard to maintain. Also now the tests should be more stable and not
have direct output to the terminal.
